File: 킬링캠프/1주 차 - 완전탐색, 백트래킹, 이분탐색/challenge/Sudoku Solver/Solution.java
class Solution {

    private static final char EMPTY = '.';

    public void solveSudoku(char[][] board) {
        backTracking(board);
    }

    private boolean backTracking(char[][] board) {
        if (checkBoardFullFill(board)) {
            return true;
        }

        for (int row = 0; row < board.length; row++) {
            for (int col = 0; col < board[row].length; col++) {
                if (board[row][col] != EMPTY) {
                    continue;
                }

                for (int num = 1; num <= 9; num++) {
                    if (!isSafe(board, row, col, num)) {
                        continue;
                    }

                    board[row][col] = (char)(num + '0');
                    if (backTracking(board)) {
                        return true;
                    }

                    board[row][col] = EMPTY;
                }

                return false;
            }
        }

        return false;
    }

    private boolean checkBoardFullFill(char[][] board) {
        for (int row = 0; row < board.length; row++) {
            for (int col = 0; col < board[row].length; col++) {
                if (board[row][col] == EMPTY) {
                    return false;
                }
            }
        }

        return true;
    }

    private boolean isSafe(char[][] board, int row, int col, int num) {
        for (int targetRow = 0; targetRow < board.length; targetRow++) {
            if (board[targetRow][col] == (char)(num + '0')) {
                return false;
            }
        }

        for (int targetCol = 0; targetCol < board.length; targetCol++) {
            if (board[row][targetCol] == (char)(num + '0')) {
                return false;
            }
        }

        int targetRow = getBaseIndex(row);
        int targetCol = getBaseIndex(col);
        for (int rowOffset = 0; rowOffset < 3; rowOffset++) {
            for (int colOffset = 0; colOffset < 3; colOffset++) {
                if (board[targetRow + rowOffset][targetCol + colOffset] == (char)(num + '0')) {
                    return false;
                }
            }
        }

        return true;
    }

    private int getBaseIndex(int target) {
        return switch(target % 3) {
            case 0 -> target;
            case 1 -> target - 1;
            case 2 -> target - 2;
            default -> -1;
        };
    } 
}

File: 킬링캠프/1주 차 - 완전탐색, 백트래킹, 이분탐색/challenge/Sudoku Solver/solution.py
class Solution:
    EMPTY_STR = "."

    def solveSudoku(self, board: List[List[str]]) -> None:
        """
        Do not return anything, modify board in-place instead.
        """
        self.back_tracking(board)
    
    def back_tracking(self, board):
        if (self.is_full_fill(board)):
            return True
        
        for row in range(len(board)):
            for col in range(len(board[row])):
                if (board[row][col] != self.EMPTY_STR):
                    continue
                
                for candi in range(1, 10):
                    if (not self.is_safe(row, col, candi, board)):
                        continue
                    
                    board[row][col] = str(candi)
                    if (self.back_tracking(board)):
                        return True

                    board[row][col] = self.EMPTY_STR

                return False

        return False
    
    def is_full_fill(self, board):
        for row in board:
            for col in row:
                if (col == self.EMPTY_STR):
                    return False
        
        return True

    def is_safe(self, target_row, target_col, target, board):
        for col in range(len(board)):
            if (col == target_col):
                continue
            
            if (board[target_row][col] == str(target)):
                return False
        
        for row in range(len(board)):
            if (row == target_row):
                continue
            
            if (board[row][target_col] == str(target)):
                return False
        
        start_row = target_row - (target_row % 3)
        start_col = target_col - (target_col % 3)
        for row in range(start_row, start_row + 3):
            for col in range(start_col, start_col + 3):
                if (row == target_row and col == target_col):
                    continue
                
                if (board[row][col] == str(target)):
                    return False
        
        return True
